### PR TITLE
Fix data lost due to incorrect sharedsnapshot

### PR DESCRIPTION
**Pull request by dreamedcheng**
7X merges upstream commit
https://github.com/postgres/postgres/commit/8a7d0701814a4e293efad22091d6f6fb441bbe1c, which uses snapshot to speed up HeapTupleSatisfiesMVCC by replacing the XID-in-progress test. However, in GPDB7, sharedsnapshot related logic didn't modified accordingly, which will result in MVCC problem when reader evaluates tuples modified by some sub-transactions, and may further cause data lost. The root cause is that sharedsnapshot didn't contain sub-transaction information of the writer gang.

We can refer to the test case of the commit to reproduce the problem.

In this commit, sub-transaction information is included in sharedsnapshot, when writer gang is updating the local sharedsnapshot.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by kainwen** [comment]
Seems the same issue here https://github.com/greenplum-db/gpdb/issues/17275

**Review by SmartKeyerror — APPROVED** [review]
LGTM.

Thanks for your contribution, I'll merge it if full pipeline passes.

**Comment by dreamedcheng** [comment]
Forget to put the test case in isolation2_schedule, added.

**Comment by haolinw** [comment]
Is it possible to add some assertion to guard no unexpected in-progress transaction status returned any more after snapshot check in `HeapTupleSatisfiesMVCC` ? For example, Assert(xidstatus != TRANSACTION_STATUS_IN_PROGRESS) after `XidInMVCCSnapshot`. Since RCA such issue could be difficult and if we have it raise a shout at unexpected scenario then could save time. cc @SmartKeyerror.

**Comment by SmartKeyerror** [comment]
> Is it possible to add some assertion to guard no unexpected in-progress transaction status returned any more after snapshot check in `HeapTupleSatisfiesMVCC` ? For example, Assert(xidstatus != TRANSACTION_STATUS_IN_PROGRESS) after `XidInMVCCSnapshot`. Since RCA such issue could be difficult and if we have it raise a shout at unexpected scenario then could save time. cc @SmartKeyerror.

It seems difficult to add such assertions, because the process that tries to update the HintBits is just read-only, it can not
verify what's right and what's wrong.

**Comment by haolinw** [comment]
> > Is it possible to add some assertion to guard no unexpected in-progress transaction status returned any more after snapshot check in `HeapTupleSatisfiesMVCC` ? For example, Assert(xidstatus != TRANSACTION_STATUS_IN_PROGRESS) after `XidInMVCCSnapshot`. Since RCA such issue could be difficult and if we have it raise a shout at unexpected scenario then could save time. cc @SmartKeyerror.
> 
> It seems difficult to add such assertions, because the process that tries to update the HintBits is just read-only, it can not verify what's right and what's wrong.

As a quick example: https://github.com/greenplum-db/gpdb/commit/54de020ed36a4d2d5fbfbb5e785d4377d03f7b20
Read only for verification purpose is enough.

**Comment by kainwen** [comment]
Mentioned in the issue page https://github.com/greenplum-db/gpdb/issues/17275#issuecomment-2038590255

The old commit https://github.com/greenplum-db/gpdb/commit/b52ca70fdfde78d4de0cd31b3a9cf7855c7a7239 seems not add any regress tests. It just added some unittests. The old commit is to fix two issues https://github.com/greenplum-db/gpdb/issues/2269 and https://github.com/greenplum-db/gpdb/issues/2284 @SmartKeyerror looks into these and by reading the code he thinks the problem does not exists nowadays due to much code refactor.

Quick sync with @ashwinstar , some concern is: subtransaction array is subtle, some cases are overflow, not overflow.

Thanks everyone working on this. Target this to 7.2, @ashwinstar might spend more time on this also.

**Comment by ashwinstar** [comment]
> Mentioned in the issue page [#17275 (comment)](https://github.com/greenplum-db/gpdb/issues/17275#issuecomment-2038590255)
> 
> The old commit [b52ca70](https://github.com/greenplum-db/gpdb/commit/b52ca70fdfde78d4de0cd31b3a9cf7855c7a7239) seems not add any regress tests. It just added some unittests. The old commit is to fix two issues #2269 and #2284 @SmartKeyerror looks into these and by reading the code he thinks the problem does not exists nowadays due to much code refactor.
> 

Given those issues don't even mention the RCA, how can we confidently say they got fixed and not affected due to this revert of functionality? Have we ran the mentioned MADLIB functions (which the modifications required to make the error happen frequently) and flow to validate the change doesn't have the problem for which the change was introduced? (I am still digging into details and once find more will post here).

Commit says "Due to lazy xid, subtransaction information can
change in the writer asynchronously with respect to the reader.  This
caused reader's subtransaction information out of date." what in new code compared to old code now makes sure Reader's subtransaction information won't be stale with this change compred to writer?

**Comment by ashwinstar** [comment]
Okay, so this PR is not touching `TransactionIdIsCurrentTransactionId()` and `IsCurrentTransactionIdForReader()` aspect of the code. That logic still remains to check transaction ID is current via writer proc only. I believe that answers by above question then why it doesn't affect the previously mentioned fixes at it related to `TransactionIdIsCurrentTransactionId()` aspect.

File: src/backend/utils/time/sharedsnapshot.c [hunk 1]
```diff
@@ -744,12 +751,21 @@ readSharedLocalSnapshot_forCursor(Snapshot snapshot, DtxContext distributedTrans
...
+	snapshot->subxcnt = dumpsnapshot->subxcnt;
 
 	memcpy(snapshot->xip, dumpsnapshot->xip, snapshot->xcnt * sizeof(TransactionId));
 
 	/* zero out the slack in the xip-array */
 	memset(snapshot->xip + snapshot->xcnt, 0, (xipEntryCount - snapshot->xcnt)*sizeof(TransactionId));
 
+	if (snapshot->subxcnt > 0)
+	{
+		memcpy(snapshot->subxip, dumpsnapshot->subxip, snapshot->subxcnt * sizeof(TransactionId));
```
Comment (ashwinstar) [inline src/backend/utils/time/sharedsnapshot.c:764]: Assert or error would be good `snapshot->subxcnt <= subXipEntryCount)`

**Review by ashwinstar — APPROVED** [review]
Overall change looks good to me. Thanks a lot!

File: src/backend/utils/time/sharedsnapshot.c (hunk 1 above)
Comment (dreamedcheng) [reply src/backend/utils/time/sharedsnapshot.c:766]: done

**Comment by kainwen** [comment]
Pushed, with one more case from the issue added. 
Thank @dreamedcheng so much for your contribution.
Thanks @ashwinstar and @SmartKeyerror for the review.

**Pull request merged**